### ISSUE TITLE
-iframework flag is not applied for pch target

### ISSUE THREAD

**Issue by drizt** [nanoant/CMakePCHCompiler#36]
Problem. Compilation error on macOS

pch compilation command:
 ```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++  -DHAVE_QT5 -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_CAST_FROM_ASCII -DQT_WIDGETS_LIB -I/Users/taurus/devel/torrent-file-editor -I/Users/taurus/devel/torrent-file-editor/test -I/Library/Frameworks/Sparkle.framework/Headers -isystem /usr/local/Qt/5.5/clang_64/lib/QtCore.framework -isystem /usr/local/Qt/5.5/clang_64/lib/QtCore.framework/Headers -isystem /usr/local/Qt/5.5/clang_64/./mkspecs/macx-clang -isystem /usr/local/Qt/5.5/clang_64/lib/QtGui.framework -isystem /usr/local/Qt/5.5/clang_64/lib/QtGui.framework/Headers -isystem /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/OpenGL.framework/Headers -isystem /usr/local/Qt/5.5/clang_64/lib/QtWidgets.framework -isystem /usr/local/Qt/5.5/clang_64/lib/QtWidgets.framework/Headers  -stdlib=libc++ -Wall -Wextra -pedantic -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.7   -std=gnu++11 -fPIC -x c++-header -o CMakeFiles/torrent-file-editor.pch.dir/pch.h.pch -c /Users/taurus/devel/torrent-file-editor/pch.h
```

But must be something link
```
clang++  -DHAVE_QT5 -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_CAST_FROM_ASCII -DQT_WIDGETS_LIB -I../ -I. -I/Library/Frameworks/Sparkle.framework/Headers -iframework /usr/local/Qt/5.5/clang_64/lib -isystem /usr/local/Qt/5.5/clang_64/lib/QtCore.framework/Headers -isystem /usr/local/Qt/5.5/clang_64/./mkspecs/macx-clang -isystem /usr/local/Qt/5.5/clang_64/lib/QtGui.framework/Headers -isystem /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/OpenGL.framework/Headers -isystem /usr/local/Qt/5.5/clang_64/lib/QtWidgets.framework/Headers -stdlib=libc++ -Wall -Wextra -pedantic -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.7   -F/Library/Frameworks  -include /Users/taurus/devel/torrent-file-editor/pch.h -fPIC -std=gnu++11 -MD -MT CMakeFiles/torrent-file-editor.dir/datewidget.cpp.o -MF CMakeFiles/torrent-file-editor.dir/datewidget.cpp.o.d -o CMakeFiles/torrent-file-editor.dir/datewidget.cpp.o -c ../datewidget.cpp
```

Second command has `-iframework /usr/local/Qt/5.5/clang_64/lib`. In first command this flag is missing. You can look error on [Travis](https://travis-ci.org/torrent-file-editor/torrent-file-editor/jobs/526358088#L1538).

CMake generator has complex rules to make includes flags. [CMake sources](https://github.com/Kitware/CMake/blob/master/Source/cmLocalGenerator.cxx#L778). CMake generator loop over includes dirs and tries to find directory where frameworks are placed. So if it see `/usr/local/Qt/5.5/clang_64/lib/QtGui.framework` it drop last path part i.e. new string will be `/usr/local/Qt/5.5/clang_64/lib` then prepend it with `-iframework`. 

On `/usr/local/Qt/5.5/clang_64/lib/QtGui.framework` input.
Output will be: `-iframework /usr/local/Qt/5.5/clang_64/lib`.

I don't know why such transformation is not applied for pch target.

**Comment by drizt** [comment]
Seems pch target is not get FRAMEWORK and INTERFACE_INCLUDE_DIRECTORIES propties.

[dump-properties.txt](https://github.com/nanoant/CMakePCHCompiler/files/3131965/dump-properties.txt)

**Comment by nanoant** [comment]
@drizt Can you please check if adding one of these (please first try one by one) in `CMakePCHCompiler.cmake` in `function(__watch_pch_last_hook ...` under `foreach(property` does the job.
Like it is mentioned there in the comment I am trying to copy most common properties, but the list is very likely incomplete.

**Comment by drizt** [comment]
It's not work. The full list of propertes:
```torrent-file-editor AUTOGEN_ORIGIN_DEPENDS = ON
torrent-file-editor AUTOMOC_COMPILER_PREDEFINES = ON
torrent-file-editor AUTOMOC_MACRO_NAMES = Q_OBJECT;Q_GADGET;Q_NAMESPACE
torrent-file-editor BINARY_DIR = /Users/taurus/devel/torrent-file-editor/debug
torrent-file-editor BINARY_DIR = /Users/taurus/devel/torrent-file-editor/debug
torrent-file-editor BUILD_WITH_INSTALL_RPATH = OFF
torrent-file-editor COMPILE_OPTIONS = -include;/Users/taurus/devel/torrent-file-editor/debug/CMakeFiles/torrent-file-editor.pch.dir/pch.h
torrent-file-editor COMPILE_OPTIONS = -include;/Users/taurus/devel/torrent-file-editor/debug/CMakeFiles/torrent-file-editor.pch.dir/pch.h
torrent-file-editor COMPILE_OPTIONS = -include;/Users/taurus/devel/torrent-file-editor/debug/CMakeFiles/torrent-file-editor.pch.dir/pch.h
torrent-file-editor CXX_STANDARD = 11
torrent-file-editor CXX_STANDARD_REQUIRED = ON
torrent-file-editor IMPORTED = FALSE
torrent-file-editor IMPORTED_GLOBAL = FALSE
torrent-file-editor INCLUDE_DIRECTORIES = /Users/taurus/devel/torrent-file-editor;/Users/taurus/devel/torrent-file-editor/debug;/Library/Frameworks/Sparkle.framework/Headers
torrent-file-editor INCLUDE_DIRECTORIES = /Users/taurus/devel/torrent-file-editor;/Users/taurus/devel/torrent-file-editor/debug;/Library/Frameworks/Sparkle.framework/Headers
torrent-file-editor INCLUDE_DIRECTORIES = /Users/taurus/devel/torrent-file-editor;/Users/taurus/devel/torrent-file-editor/debug;/Library/Frameworks/Sparkle.framework/Headers
torrent-file-editor INSTALL_RPATH = 
torrent-file-editor INSTALL_RPATH_USE_LINK_PATH = OFF
torrent-file-editor INTERFACE_LINK_LIBRARIES = Qt5::Core;Qt5::Gui;Qt5::Widgets;/Library/Frameworks/Sparkle.framework;-framework AppKit
torrent-file-editor LINK_LIBRARIES = Qt5::Core;Qt5::Gui;Qt5::Widgets;/Library/Frameworks/Sparkle.framework;-framework AppKit
torrent-file-editor MACOSX_BUNDLE = ON
torrent-file-editor MANUALLY_ADDED_DEPENDENCIES = torrent-file-editor.pch;update_version
torrent-file-editor NAME = torrent-file-editor
torrent-file-editor OSX_ARCHITECTURES = 
torrent-file-editor OUTPUT_NAME = Torrent File Editor
torrent-file-editor SKIP_BUILD_RPATH = OFF
torrent-file-editor SOURCES = /Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_ar.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_bn.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_cs.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_de.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_en.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_es.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_fi.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_fr.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_he.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_ko.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_hu.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_id.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_it.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_ja.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_nl.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_pl.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_pt.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_ro.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_ru.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_tr.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_vi.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_zh_CN.qm;/Users/taurus/devel/torrent-file-editor/debug/torrentfileeditor_zh_TW.qm;/Users/taurus/devel/torrent-file-editor/application.h;/Users/taurus/devel/torrent-file-editor/mainwindow.h;/Users/taurus/devel/torrent-file-editor/datewidget.h;/Users/taurus/devel/torrent-file-editor/lineeditwidget.h;/Users/taurus/devel/torrent-file-editor/urledit.h;/Users/taurus/devel/torrent-file-editor/folderedit.h;/Users/taurus/devel/torrent-file-editor/aboutdlg.h;/Users/taurus/devel/torrent-file-editor/bencodemodel.h;/Users/taurus/devel/torrent-file-editor/bencodedelegate.h;/Users/taurus/devel/torrent-file-editor/tableview.h;/Users/taurus/devel/torrent-file-editor/treeview.h;/Users/taurus/devel/torrent-file-editor/combobox.h;/Users/taurus/devel/torrent-file-editor/searchdlg.h;/Users/taurus/devel/torrent-file-editor/plaintextedit.h;/Users/taurus/devel/torrent-file-editor/bencode.h;/Users/taurus/devel/torrent-file-editor/abstracttreemodel.h;/Users/taurus/devel/torrent-file-editor/abstracttreenode.h;/Users/taurus/devel/torrent-file-editor/proxystyle.h;/Users/taurus/devel/torrent-file-editor/debug/config.h;/Users/taurus/devel/torrent-file-editor/application.cpp;/Users/taurus/devel/torrent-file-editor/main.cpp;/Users/taurus/devel/torrent-file-editor/mainwindow.cpp;/Users/taurus/devel/torrent-file-editor/bencode.cpp;/Users/taurus/devel/torrent-file-editor/datewidget.cpp;/Users/taurus/devel/torrent-file-editor/lineeditwidget.cpp;/Users/taurus/devel/torrent-file-editor/urledit.cpp;/Users/taurus/devel/torrent-file-editor/folderedit.cpp;/Users/taurus/devel/torrent-file-editor/aboutdlg.cpp;/Users/taurus/devel/torrent-file-editor/bencodemodel.cpp;/Users/taurus/devel/torrent-file-editor/bencodedelegate.cpp;/Users/taurus/devel/torrent-file-editor/proxystyle.cpp;/Users/taurus/devel/torrent-file-editor/tableview.cpp;/Users/taurus/devel/torrent-file-editor/treeview.cpp;/Users/taurus/devel/torrent-file-editor/combobox.cpp;/Users/taurus/devel/torrent-file-editor/searchdlg.cpp;/Users/taurus/devel/torrent-file-editor/plaintextedit.cpp;/Users/taurus/devel/torrent-file-editor/cocoainitializer.mm;/Users/taurus/devel/torrent-file-editor/sparkleautoupdater.mm;/Users/taurus/devel/torrent-file-editor/debug/moc_application.cpp;/Users/taurus/devel/torrent-file-editor/debug/moc_mainwindow.cpp;/Users/taurus/devel/torrent-file-editor/debug/moc_datewidget.cpp;/Users/taurus/devel/torrent-file-editor/debug/moc_lineeditwidget.cpp;/Users/taurus/devel/torrent-file-editor/debug/moc_urledit.cpp;/Users/taurus/devel/torrent-file-editor/debug/moc_folderedit.cpp;/Users/taurus/devel/torrent-file-editor/debug/moc_aboutdlg.cpp;/Users/taurus/devel/torrent-file-editor/debug/moc_bencodemodel.cpp;/Users/taurus/devel/torrent-file-editor/debug/moc_bencodedelegate.cpp;/Users/taurus/devel/torrent-file-editor/debug/moc_tableview.cpp;/Users/taurus/devel/torrent-file-editor/debug/moc_treeview.cpp;/Users/taurus/devel/torrent-file-editor/debug/moc_combobox.cpp;/Users/taurus/devel/torrent-file-editor/debug/moc_searchdlg.cpp;/Users/taurus/devel/torrent-file-editor/debug/moc_plaintextedit.cpp;/Users/taurus/devel/torrent-file-editor/debug/qrc_translations.cpp;/Users/taurus/devel/torrent-file-editor/debug/qrc_resources.cpp;/Users/taurus/devel/torrent-file-editor/debug/ui_mainwindow.h;/Users/taurus/devel/torrent-file-editor/debug/ui_aboutdlg.h;/Users/taurus/devel/torrent-file-editor/debug/ui_searchdlg.h
torrent-file-editor SOURCE_DIR = /Users/taurus/devel/torrent-file-editor
torrent-file-editor SOURCE_DIR = /Users/taurus/devel/torrent-file-editor
torrent-file-editor TYPE = EXECUTABLE
torrent-file-editor TYPE = EXECUTABLE
torrent-file-editor WIN32_EXECUTABLE = ON
```

**Comment by nanoant** [comment]
@drizt Okay, I will try to reproduce it, but can you give me full specs of the environment OS, CMake ver, Compiler Ver and minimal CMakeLists.txt example?

**Comment by drizt** [comment]
[pchtest.tar.gz](https://github.com/nanoant/CMakePCHCompiler/files/3134668/pchtest.tar.gz)

cmake version 3.14.3  
Apple LLVM version 10.0.0 (clang-1000.11.45.5)  
Qt 5.5.1
macOS 10.14
SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk  
MACOSX_DEPLOYMENT_TARGET=10.7

**Comment by drizt** [comment]
```
[taurus@Ivans-Air build]$ cmake -DCMAKE_BUILD_TYPE=Debug ..
-- The CXX compiler identification is AppleClang 10.0.0.10001145
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/taurus/devel/pchtest/build
[taurus@Ivans-Air build]$ make
/usr/local/Cellar/cmake/3.14.3/bin/cmake -S/Users/taurus/devel/pchtest -B/Users/taurus/devel/pchtest/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/local/Cellar/cmake/3.14.3/bin/cmake -E cmake_progress_start /Users/taurus/devel/pchtest/build/CMakeFiles /Users/taurus/devel/pchtest/build/CMakeFiles/progress.marks
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/Makefile2 all
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/pchtest.pch_autogen.dir/build.make CMakeFiles/pchtest.pch_autogen.dir/depend
cd /Users/taurus/devel/pchtest/build && /usr/local/Cellar/cmake/3.14.3/bin/cmake -E cmake_depends "Unix Makefiles" /Users/taurus/devel/pchtest /Users/taurus/devel/pchtest /Users/taurus/devel/pchtest/build /Users/taurus/devel/pchtest/build /Users/taurus/devel/pchtest/build/CMakeFiles/pchtest.pch_autogen.dir/DependInfo.cmake --color=
Scanning dependencies of target pchtest.pch_autogen
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/pchtest.pch_autogen.dir/build.make CMakeFiles/pchtest.pch_autogen.dir/build
[ 14%] Automatic MOC for target pchtest.pch
/usr/local/Cellar/cmake/3.14.3/bin/cmake -E cmake_autogen /Users/taurus/devel/pchtest/build/CMakeFiles/pchtest.pch_autogen.dir/AutogenInfo.cmake Debug
[ 14%] Built target pchtest.pch_autogen
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/pchtest.pch.dir/build.make CMakeFiles/pchtest.pch.dir/depend
cd /Users/taurus/devel/pchtest/build && /usr/local/Cellar/cmake/3.14.3/bin/cmake -E cmake_depends "Unix Makefiles" /Users/taurus/devel/pchtest /Users/taurus/devel/pchtest /Users/taurus/devel/pchtest/build /Users/taurus/devel/pchtest/build /Users/taurus/devel/pchtest/build/CMakeFiles/pchtest.pch.dir/DependInfo.cmake --color=
Scanning dependencies of target pchtest.pch
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/pchtest.pch.dir/build.make CMakeFiles/pchtest.pch.dir/build
[ 28%] Building CXX object CMakeFiles/pchtest.pch.dir/pchtest.pch_autogen/mocs_compilation.cpp.o
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++  -DQT_CORE_LIB -I/Users/taurus/devel/pchtest/build -I/Users/taurus/devel/pchtest -I/Users/taurus/devel/pchtest/build/pchtest.pch_autogen/include -iframework /usr/local/Qt/5.5/clang_64/lib -isystem /usr/local/Qt/5.5/clang_64/lib/QtCore.framework/Headers -isystem /usr/local/Qt/5.5/clang_64/./mkspecs/macx-clang  -stdlib=libc++ -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.7   -std=gnu++11 -fPIC -std=gnu++11 -o CMakeFiles/pchtest.pch.dir/pchtest.pch_autogen/mocs_compilation.cpp.o -c /Users/taurus/devel/pchtest/build/pchtest.pch_autogen/mocs_compilation.cpp
[ 42%] Building CXXPCH object CMakeFiles/pchtest.pch.dir/pch.h.pch
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++  -DQT_CORE_LIB -I/Users/taurus/devel/pchtest/build -I/Users/taurus/devel/pchtest -I/Users/taurus/devel/pchtest/build/pchtest.pch_autogen/include -isystem /usr/local/Qt/5.5/clang_64/lib/QtCore.framework -isystem /usr/local/Qt/5.5/clang_64/lib/QtCore.framework/Headers -isystem /usr/local/Qt/5.5/clang_64/./mkspecs/macx-clang  -stdlib=libc++ -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.7   -std=gnu++11 -fPIC -x c++-header -o CMakeFiles/pchtest.pch.dir/pch.h.pch -c /Users/taurus/devel/pchtest/pch.h
In file included from /Users/taurus/devel/pchtest/pch.h:1:
In file included from /usr/local/Qt/5.5/clang_64/lib/QtCore.framework/Headers/QCoreApplication:1:
/usr/local/Qt/5.5/clang_64/lib/QtCore.framework/Headers/qcoreapplication.h:37:10: fatal error: 'QtCore/qglobal.h' file not found
#include <QtCore/qglobal.h>
         ^~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [CMakeFiles/pchtest.pch.dir/pch.h.pch] Error 1
make[1]: *** [CMakeFiles/pchtest.pch.dir/all] Error 2
make: *** [all] Error 2
```
